File: generator_web_app/src/modules/api/images.ts
// API methods for managing image objects
// Author: Michael B. Lance

// ---------------------------------------------------------------------------------------------------------------------------

import {
  Image,
  Prediction,
  type ImageIntf,
  PredictionCrop,
  type PredictionCropIntf,
  type PredictionIntf,
  type imageRecords,
  Annotation,
  type AnnotationIntf,
} from "@/types/generatorobjects";
import { ApiError } from "@/modules/api/errors";
import { api_url } from "@/modules/api/apiV1Methods";

// ---------------------------------------------------------------------------------------------------------------------------
//GET

export async function getImagePredictions(
  image_id: string,
): Promise<imageRecords["predictions"]> {
  const response = await fetch(`${api_url}/images/${image_id}/predictions`, {
    method: "GET",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();

  let predictions: imageRecords["predictions"] = {};
  for (const pred of resp)
    predictions[pred.uuid] = new Prediction(pred as PredictionIntf);

  return predictions;
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function getImageAnnotations(
  image_id: string,
): Promise<imageRecords["annotations"]> {
  const response = await fetch(`${api_url}/images/${image_id}/annotations`, {
    method: "GET",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();

  let annotations: imageRecords["annotations"] = {};
  for (const annot of resp)
    annotations[annot.uuid] = new Annotation(annot as AnnotationIntf);

  return annotations;
}

// ---------------------------------------------------------------------------------------------------------------------------
//POST

interface createImageOptions {
  survey_id: string | number;
  herd_unit_id: string | number;
  name: string;
  img_key: string;
  image_length_px: number;
  image_width_px: number;
}

export async function createImage(options: createImageOptions): Promise<Image> {
  const response = await fetch(`${api_url}/images`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
  });

  if (!response.ok) throw new ApiError(await response.json());

  return new Image((await response.json()) as ImageIntf);
}

// ---------------------------------------------------------------------------------------------------------------------------

interface presignedPutOptions {
  upload_id: string;
  part_number: number;
  image_id: number | string;
  chunk_size: number;
  chunk_md5: string;
}

export async function createImagePresignedPut(
  options: presignedPutOptions,
): Promise<string> {
  const response = await fetch(`${api_url}/images/presigned-put-url`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
  });

  if (!response.ok) throw new ApiError(await response.json());

  return await response.text();
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function createMultiPartUpload(
  image_key: string,
): Promise<string> {
  const response = await fetch(`${api_url}/images/create-multipart-upload`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      image_key: image_key,
    }),
  });
  if (!response.ok) throw new ApiError(await response.json());

  return await response.text();
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function completeMultiPartUpload(
  image_key: string,
  parts: any[],
  upload_id: string,
): Promise<string> {
  const response = await fetch(`${api_url}/images/complete-multipart-upload`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      image_key: image_key,
      parts: parts,
      upload_id: upload_id,
    }),
  });
  if (!response.ok) throw new ApiError(await response.json());

  return (await response.json()) as string;
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function abortMultipartUpload(
  image_key: string,
  upload_id: string,
): Promise<string> {
  const response = await fetch(`${api_url}/images/abort-multipart-upload`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      image_key: image_key,
      upload_id: upload_id,
    }),
  });
  if (!response.ok) throw new ApiError(await response.json());

  return (await response.json()) as string;
}

// ---------------------------------------------------------------------------------------------------------------------------

interface CreatePredCropsOptions {
  image_id: string;
  prediction_id: string[];
}

export async function createPredCrops(
  options: CreatePredCropsOptions,
): Promise<PredictionCrop[]> {
  const response = await fetch(`${api_url}/images/prediction_crops`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      image_id: options.image_id,
      prediction_id: options.prediction_id,
    }),
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();
  const pred_crops: PredictionCrop[] = [];

  for (const row of resp) {
    const predCrop = row as PredictionCropIntf;

    pred_crops.push(
      new PredictionCrop(
        predCrop,
        `${api_url}/images/prediction_crops/${predCrop.uuid}`,
      ),
    );
  }

  return pred_crops;
}

// ---------------------------------------------------------------------------------------------------------------------------
// PATCH

interface updateImageOptions {
  name?: string;
  herd_unit_id?: number;
  survey_id?: number;
  img_key?: string;
  image_length_px?: number;
  image_width_px?: number;
  area?: number;
  viewshed_polygon?: number[][];
  has_detection?: boolean;
  dem_name?: string;
  bbox_wsen?: number[];
  opened_by_user_id?: number;
}

export async function updateImage(
  image_id: string,
  options: updateImageOptions,
): Promise<Image> {
  const response = await fetch(`${api_url}/images/${image_id}`, {
    method: "PATCH",

    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
  });

  if (!response.ok) throw new ApiError(await response.json());

  return new Image((await response.json()) as ImageIntf);
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function closeUserImages(): Promise<boolean> {
  const response = await fetch(`${api_url}/images/close-user-images`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (!response.ok) throw new ApiError(await response.json());

  return true;
}

// ---------------------------------------------------------------------------------------------------------------------------
// DELETE

export async function deleteImage(image_id: string): Promise<boolean> {
  const response = await fetch(`${api_url}/images/${image_id}`, {
    method: "DELETE",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  return true;
}

File: airial_api/app/routers/images.py
# Endpoints for managing images in the API
# Author: Michael B. Lance

# ---------------------------------------------------------------------------------------------------------------------------

from uuid import UUID

import cv2
from flask import Blueprint, Response, abort, current_app
from flask_login import current_user, login_required
from flask_pydantic import validate
from app.decorators import permission_required, roles_required

from app.extensions import base, cache, s3
from crop_generator import create_subcrop
from database.errors import ObjectNotFound
from database.object_models.core import (
    CreateImageReq,
    CreatePredictionCropReq,
    CreatePresignedPutReq,
    PredictionQuery,
    UpdateImageReq,
)
from database.object_models.core.images import (
    AbortMulitPartUploadReq,
    CompleteMultiPartUploadReq,
    CreateImageMultiPartUploadReq,
    PresignedImageGetReq,
)

imageBp = Blueprint("images", __name__, url_prefix="/api/v1/images")

# ---------------------------------------------------------------------------------------------------------------------------
# GET


@imageBp.get("/<string:image_id>")
@login_required
@permission_required("access")
def get_by_id(image_id: str):
    """
    Request an image object from the database using its UUID.
    ---
    parameters:
      - name: image_id
            in: path
            type: string
            required: true
    responses:
      200:
            description: The requested image was found.
      400:
            description: Invalid UUID format provided.
      404:
            description: No image record found for the provided ID.
    """
    return base.get_image(UUID(image_id)).to_dict(), 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@permission_required("access")
@imageBp.get("/prediction_crops/<string:prediction_id>")
def get_prediction_crop(prediction_id: str):
    """ """
    crop = cache.get(prediction_id)
    if crop is None:
        raise ObjectNotFound("prediction_crop", prediction_id)

    _, encoded_image = cv2.imencode(".webp", crop)
    return Response(encoded_image.tobytes(), mimetype="image/webp"), 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.get("/<string:image_id>/crops")
@login_required
def get_crops(image_id: str):
    """
    Retrieve crops for a specific image
    ---
    responses:
      200:
            description: List of crops.
      404:
            description: No crops found.
      500:
            description: Database error.
    """

    crops = base.get_image_crops(UUID(image_id))

    return [crop.to_dict() for crop in crops], 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.get("/<string:image_id>/predictions")
@login_required
@permission_required("access")
def get_predictions(image_id: str):
    """
    Retrieve all predictions associated with an image.
    ---
    parameters:
            - name: image_id
            in: path
            type: string
            required: true
    responses:
            200:
                    description: List of predictions.
            400:
                    description: Invalid UUID format.
            404:
                    description: No predictions found.
            500:
                    description: Database error.
    """

    predictions = base.get_image_predictions(UUID(image_id))

    return [pred.to_dict() for pred in predictions], 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.get("/<string:image_id>/annotations")
@login_required
@permission_required("access")
def get_annotations(image_id: str):
    """
    Retrieve all annotations associated with an image.
    ---
    parameters:
            - name: image_id
              in: path
              type: string
              required: true
    responses:
            200:
                    description: Annotations belonging to a given image.
            401:
                    description: User not authorized to perform this action.
            400:
                    description: Malformed request.
            404:
                    description: Image not found.
            500:
                    description: Unexpected error.

    """

    annotations = base.get_image_annotations(UUID(image_id))

    return [annot.to_dict() for annot in annotations], 200


# ---------------------------------------------------------------------------------------------------------------------------
# POST


@imageBp.post("")
@login_required
@permission_required("access")
@validate()
def create(body: CreateImageReq):
    """
    Create a new image object.
    ---
    parameters:
            - name: create image request
            in: body
            type: CreateImageReq
            required: true
    responses:
            201:
                    description: Created successfully.
            409:
                    description: Image already exists (Unique Violation).
            400:
            500:
                    description: Database error.
    """
    return base.create_image(body).to_dict(), 201


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/presigned-get-url")
@login_required
@permission_required("access")
@validate()
def create_presigned_get(body: PresignedImageGetReq):
    """
    Generate a presigned GET URL for an image.
    ---
    responses:
            201:
                    description: Presigned URL generated.
            400:
                    description: Invalid ID format.
            404:
                    description: Image record not found.
            500:
                    description: Storage or database error.
    """

    image = base.get_image(body.image_id)

    if not image:
        raise ObjectNotFound("image", str(body.image_id))

    response = s3.generate_presigned_url(
        "get_object",
        Params={"Bucket": current_app.config["BUCKET_NAME"], "Key": image.img_key},
        ExpiresIn=body.image_id,
    )

    return response, 201


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/presigned-put-url")
@roles_required("admin")
@validate()
@login_required
@permission_required("access")
def create_chunk_presigned_put(body: CreatePresignedPutReq):
    """
    Generates a pre-signed URL for a single file chunk (UploadPart).
    ---
    responses:
            201:
                    description: URL generated.
            400:
                    description: Missing data or invalid UUID.
            404:
                    description: Image not found.
            500:
                    description: Storage/Database error.
    """
    data = body.model_dump()

    image_id = (
        UUID(data["image_id"])
        if isinstance(data["image_id"], str)
        else data["image_id"]
    )
    image = base._get_image(image_id)

    if image is None:
        abort(404, "Image not found")

    response = s3.generate_presigned_url(
        ClientMethod="upload_part",
        Params={
            "Bucket": current_app.config["BUCKET_NAME"],
            "Key": image.img_key,
            "UploadId": data["upload_id"],
            "PartNumber": data["part_number"],
            "ContentLength": data["chunk_size"],
            "ContentMD5": data["chunk_md5"],
        },
        ExpiresIn=3600,
    )

    return response, 201


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/create-multipart-upload")
@login_required
@permission_required("access")
@validate()
def create_multipart_upload(body: CreateImageMultiPartUploadReq):
    """
    Initiates a new multipart upload.
    ---
    responses:
            201:
                    description: Multipart upload started.
            500:
                    description: system error.
    """
    response = s3.create_multipart_upload(
        Bucket=current_app.config["BUCKET_NAME"],
        Key=body.image_key,
        ContentType="image/jpeg",
    )

    return response["UploadId"], 201


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/complete-multipart-upload")
@login_required
@permission_required("access")
@validate()
def complete_upload(body: CompleteMultiPartUploadReq):
    """
    Completes a new multipart upload.
    ---
    responses:
            201:
                    description: Multipart upload finished.
            500:
                    description: S3 or system error.
    """

    response = s3.complete_multipart_upload(
        Bucket=current_app.config["BUCKET_NAME"],
        Key=body.image_key,
        MultipartUpload={"Parts": body.parts},
        UploadId=body.upload_id,
    )

    return response, 201


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/abort-multipart-upload")
@login_required
@permission_required("access")
@validate()
def abort_upload(body: AbortMulitPartUploadReq):
    """
    Aborts a multipart upload.
    ---
    responses:
            200:
                    description: Multipart upload aborted.
            500:
                    description: S3 or system error.
    """

    response = s3.abort_multipart_upload(
        Bucket=current_app.config["BUCKET_NAME"],
        Key=body.image_key,
        UploadId=body.upload_id,
    )

    return response, 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.post("/prediction_crops")
@login_required
@permission_required("acecss")
@validate()
def get_prediciton(body: CreatePredictionCropReq):
    """"""
    image = base.get_image(body.image_id)
    predictions = base.get_predictions(
        PredictionQuery(prediction_id=body.prediction_id)
    )

    image_data = cache.get(image.uuid)

    if not image_data:
        image_data = s3.get_object(
            Bucket=current_app.config["BUCKET_NAME"], Key=image.img_key
        )["Body"].read()

        cache.set(image.uuid, image_data, 500)

    image.set_image(image_data)
    pred_crops = create_subcrop(image, predictions)

    [cache.set(crop.uuid, crop.get_image(), 3600) for crop in pred_crops]

    return [crop.to_dict() for crop in pred_crops]


# ---------------------------------------------------------------------------------------------------------------------------
# PATCH


@imageBp.patch("/<string:image_id>")
@login_required
@permission_required("acecss")
@validate()
def update(body: UpdateImageReq, image_id: str):
    """
    Update image object in the database.
    ---
    paramete:qrsrs:
            - name: image_id
            in: path
            type: string
            required: true
    responses:
            200:
                    description: Image updated successfully.
            400:
                    description: Invalid UUID.
            404:
                    description: Image not found.
            500:
                    description: Database error.
    """

    return base.update_image(UUID(image_id), body).to_dict(), 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@imageBp.patch("/close-user-images")
@login_required
@permission_required("acess")
def close_user_images():
    """ """

    base.close_user_images(current_user.user_id)

    return "", 204


# ---------------------------------------------------------------------------------------------------------------------------
# DELETE


@imageBp.delete("/<string:image_id>")
@login_required
@permission_required("access")
def delete_image(image_id: str):
    """ """

    image = base.get_image(UUID(image_id))

    s3.delete_object(Bucket=current_app.config["BUCKET_NAME"], Key=image.img_key)

    base.delete_image(UUID(image_id))

    return "", 204
